lib/recursive-methods.rb

```diff
@@ -1,49 +1,100 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0 
+        raise ArgumentError.new("Number must be greater than zero!")
+    elsif n == 1
+        return n
+    elsif n == 0
+        return 1
+    else
+        return n * factorial(n - 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if !s || s.length == 0
+        return s
+    else
+        return reverse(s[1..-1]) << s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
```

Comment:
⚠️   This is not done in place, as you're creating a new array with each iteration.

Think about it if you added this to the method signature

```suggestion
# Time complexity: O(n)
# Space complexity: O(n)
def reverse_inplace(s, first=0, last=s.length - 1)
```